homeassistant/components/toon/__init__.py

```diff
@@ -1,142 +1,92 @@
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component))
 
+    return True
+
+
+class ToonEntity(Entity):
+    """Defines a base Toon entity."""
+
+    def __init__(self, toon, name: str, icon: str) -> None:
+        """Initialize the Toon entity."""
+        self._name = name
+        self._state = None
+        self._icon = icon
         self.toon = toon
-        self.gas = gas
-        self.solar = solar
-        self.data = {}
-
-        self.last_update = datetime.min
-        self.update()
-
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self):
-        """Update Toon data."""
-        self.last_update = datetime.now()
-
-        self.data['power_current'] = self.toon.power.value
-        self.data['power_today'] = round(
-            (float(self.toon.power.daily_usage) +
-             float(self.toon.power.daily_usage_low)) / 1000, 2)
-        self.data['temp'] = self.toon.temperature
-
-        if self.toon.thermostat_state:
-            self.data['state'] = self.toon.thermostat_state.name
-        else:
-            self.data['state'] = 'Manual'
-
-        self.data['setpoint'] = float(
-            self.toon.thermostat_info.current_set_point) / 100
-        self.data['gas_current'] = self.toon.gas.value
-        self.data['gas_today'] = round(float(self.toon.gas.daily_usage) /
-                                       1000, 2)
-
-        for plug in self.toon.smartplugs:
-            self.data[plug.name] = {
-                'current_power': plug.current_usage,
-                'today_energy': round(float(plug.daily_usage) / 1000, 2),
-                'current_state': plug.current_state,
-                'is_connected': plug.is_connected,
-            }
 
-        self.data['solar_maximum'] = self.toon.solar.maximum
-        self.data['solar_produced'] = self.toon.solar.produced
-        self.data['solar_value'] = self.toon.solar.value
-        self.data['solar_average_produced'] = self.toon.solar.average_produced
-        self.data['solar_meter_reading_low_produced'] = \
-            self.toon.solar.meter_reading_low_produced
-        self.data['solar_meter_reading_produced'] = \
-            self.toon.solar.meter_reading_produced
-        self.data['solar_daily_cost_produced'] = \
-            self.toon.solar.daily_cost_produced
-
-        for detector in self.toon.smokedetectors:
-            value = '{}_smoke_detector'.format(detector.name)
-            self.data[value] = {
-                'smoke_detector': detector.battery_level,
-                'device_type': detector.device_type,
-                'is_connected': detector.is_connected,
-                'last_connected_change': detector.last_connected_change,
+    @property
+    def device_info(self) -> Dict[str, Any]:
+        """Return device information about this entity."""
+        return {
+            'identifiers': {
+                (DOMAIN, self.toon.agreement.id)
```

Comment:
This gives all Toon entities the same unique ID. Can you only link one device with Toon ?

Reply:
There is no information (sensible) device on the other devices and all data is provided by the Toon. I've considered breaking this out in a first attempt, but not sure if that makes things better.

e.g., I added a Power meter device and a solar device... but what qualifies as power or solar? Which sensor goes where? The thermostat could be split... nevertheless, it is that same thermostat that hooks up to the power meter.

So in reality, it is a single device that gathers data (at least, that is how I eventually looked at it).

If lights and switches from the Toon would be added, I would definitely break them out.

Reply:
Well it's lights and switches that got me thinking about this. What kind of ID would you give a light?

Reply:
Would you actually expose lights and switches via Toon? 

Toon supports Hue and Z-wave devices - both of which are probably better set up via their native HA integration. Or is there a use-case for letting those devices link to HA via Toon?

Reply:
Lights probably not since HA supports hue nativelly, but toon has z wave and not everyone has a zwave adapter. In that case switches and smoke alarms can be handled through toon.

Reply:
All devices expose their uuid actually.

Reply:
@hmmbob @costastf Thanks for your responses, nevertheless, my response/post about how to divide platforms into devices hits a bigger subject: What is considered a device?

I did not mean to discuss the exposure of lights, switches or smoke detectors within this component, this was already set to a possible future version/addition.

Reply:
@balloob I tried to visualize the Toon connections and equipment involved:

https://www.draw.io/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1#G18t2uyTYO7g3Kn4xj55J--TLEc-K59495

What would your suggestion be?

Reply:
FYI, draw.io diagram is not accessible.

If linked devices get their own unique ID, I guess it's fine as we can guarantee uniqueness.

What about 2 Toon thermostats? Is that possible?

Reply:
Yes it is, but they will have a unique agreement id

Reply:
Then this is fine as it is.

Reply:
Check, still, I'm going to look into restructuring the devices as per updated description home-assistant/developers.home-assistant#182

Reply:
This has been addressed  and now reflects the actual real world:

![image](https://user-images.githubusercontent.com/195327/53289737-2cf11880-379a-11e9-837a-f301d7f461c1.png)

The "Gas Meter" will either be hooked up to the "Electricity Meter" or the "Meter Adapter" based on the fact if it is "smart" or not.

Reply:
Comments addressed in the current version.